README.md
# 决策表引擎 dmn-engine

[![Build Status](https://travis-ci.org/supermy/dmn-rest.svg?branch=master)](https://github.com/supermy/dmn-rest)

## 简介 
* 决策表引擎dmn engine；rest api 方式提供服务； 通过集群提升 QPS 能力。



## 特点
* Rest Api 方式提供服务；
* 在线发布与更新 dmn table;
* 支持 drd;

## 适用场景

1.精准获客规则定义；
2.渠道投放策略规则定义；


## 优势：

*  即开及用，在线发布维护决策模型； 

## 劣势：

*  后台由关系 DB 支撑，需要 DB 集群与 WEB 集群提升 QPS 能力    


## 决策列表

    http://127.0.0.1:8080/file-browser.html



## 快速试用
  发布 dmn table 脚本

```

mvn spring-boot:run

curl -i -X POST -H 'Cookie: JSESSIONID=kjcddrwwo1zo1ejsx4bbmek0u' -H 'Connection: keep-alive' -F "data=@/服务器文件路径/example-dmn-rest/target/classes/dmn/decision-model.dmn;tenant-id=1;deployment-source=process-application;deployment-name=abc" http://127.0.0.1:8080/rest/deployment/create

curl -i -X POST -H 'Cookie: JSESSIONID=kjcddrwwo1zo1ejsx4bbmek0u' -H 'Connection: keep-alive' -F "data=@/Users/moyong/project/env-rule/dmn-rest/src/main/resources/decision-model.dmn;tenant-id=1;deployment-source=process-application;deployment-name=abc" http://127.0.0.1:8080/rest/deployment/create

```

  执行 dmn table 脚本

```

curl -i -X POST -H 'Cookie: JSESSIONID=kjcddrwwo1zo1ejsx4bbmek0u' -H 'Connection: keep-alive ;Content-Type:application/jsons' http://127.0.0.1:8080/rest/engine/default/rest/decision-definition/key/checkOrder/evaluate
        -d "{"variables" : 
              {"status" : { "value" : "silver", "type" : "String" },
              "sum" : { "value" : 900, "type" : "Integer" }}
              }" 

```


## 支持的 api


```
    决策表列表
    GET /rest/decision-definition
    决策表数量
    GET /rest/decision-definition/count
    单个决策表
    GET /rest/decision-definition/{id} 
    单个决策表通过 key
    GET /rest/decision-definition/key/{key} 
    租户
    GET /rest/decision-definition/key/{key}/tenant-id/{tenant-id}
    通过 ID 获取XML格式决策表
    GET /rest/decision-definition/{id}/xml 
    通过 KEY 获取XML 格式决策表
    GET /rest/decision-definition/key/{key}/xml 
    获取租户
    GET /rest/decision-definition/key/{key}/tenant-id/{tenant-id}/xml
    获取图标描述
    GET /rest/decision-definition/{id}/diagram 
    
    GET /rest/decision-definition/key/{key}/diagram 
    
    GET /rest/decision-definition/key/{key}/tenant-id/{tenant-id}/diagram
    启动决策表进程
    POST /rest/decision-definition/{id}/evaluate 
    启动决策表进程
    POST /rest/decision-definition/key/{key}/evaluate 
    启动决策表进程
    POST /rest/decision-definition/key/{key}/tenant-id/{tenant-id}/evaluate
    
    Update history time to live
    PUT /rest/decision-definition/{id}/history-time-to-live 
    
    PUT /rest/decision-definition/key/{key}/history-time-to-live 
    
    PUT /rest/decision-definition/key/{key}/tenant-id/{tenant-id}/history-time-to-live
    
```

## 数据库配置，如测试使用，可注销数据库配置，默认使用 h2嵌入式数据库

```

    camunda.bpm:
      metrics.enabled: false
      history-level: auto
      database:
        schema-update: true
        type: mysql
    
    spring.datasource:
      url: jdbc:mysql://localhost:3306/camunda?useSSL=false
      username: camunda
      password: camunda
      driver-class-name: com.mysql.jdbc.Driver
    
    server:
      port: 8081
     
```

  
## 配合 dmn-js 实现 可视化编辑规则，进行决策表编辑，及时发布；所见及所得；

```
     
```


CHANGELOG.md
## 20170828
* 增加页面访问资源 http://127.0.0.1:8080/  对应 src/main/resources/static/目录
* 列表、下载、查看完成 http://127.0.0.1:8080/file-browser.html



src/test/java/org/camunda/bpm/spring/boot/example/dmn/rest/DmnRestApplicationTest.java
package org.camunda.bpm.spring.boot.example.dmn.rest;

import static org.assertj.core.api.Assertions.assertThat;

import org.camunda.bpm.engine.RepositoryService;
import org.camunda.bpm.engine.impl.util.json.JSONArray;
import org.camunda.bpm.engine.repository.DecisionDefinition;
import org.camunda.bpm.spring.boot.starter.property.CamundaBpmProperties;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;

@RunWith(SpringRunner.class)
@SpringBootTest(classes = { DmnRestApplication.class }, webEnvironment = WebEnvironment.RANDOM_PORT)
public class DmnRestApplicationTest {

  private static final String CHECK_ORDER = "checkOrder";

  @Autowired
  private TestRestTemplate testRestTemplate;

  @Autowired
  private CamundaBpmProperties camundaBpmProperties;

  @Autowired
  private RepositoryService repositoryService;

  @Test
  public void deploys_orderDiscount_dmn() {
    final DecisionDefinition definition = repositoryService.createDecisionDefinitionQuery().decisionDefinitionKey(CHECK_ORDER).singleResult();
    assertThat(definition).isNotNull();
  }

  @Test
  public void evaluate_checkOrder() throws InterruptedException {
    //输入数据格式可以是 json
    String JSONInput = "{\n" + "  \"variables\" : {\n" + "    \"status\" : { \"value\" : \"silver\", \"type\" : \"String\" },\n"
        + "    \"sum\" : { \"value\" : 900, \"type\" : \"Integer\" }\n" + "  }\n" + "}\n";

    HttpHeaders headers = new HttpHeaders();
    headers.setContentType(MediaType.APPLICATION_JSON);

    HttpEntity<String> request = new HttpEntity<String>(JSONInput, headers);

    final String check = testRestTemplate.postForObject("/rest/engine/{engineName}/decision-definition/key/{key}/evaluate", request, String.class,
        camundaBpmProperties.getProcessEngineName(), CHECK_ORDER);

    System.out.println(check);
    System.out.println(camundaBpmProperties.getProcessEngineName());

    assertThat(new JSONArray(check).getJSONObject(0).getJSONObject("result").getString("value")).isEqualTo("ok");

  }

}


src/main/java/logviewer/domain/FileNode.java
package logviewer.domain;

/**
 * Created by moyong on 16/10/28.
 */

import org.codehaus.jackson.annotate.JsonAutoDetect;
import org.codehaus.jackson.map.annotate.JsonSerialize;

import java.util.Date;
import java.util.List;

@JsonAutoDetect
@JsonSerialize(include = JsonSerialize.Inclusion.NON_NULL)
public class FileNode {

    private String text;
    public String getText() { return text; }
    public void setText(String text) { this.text = text; }

    private boolean leaf;
    public boolean isLeaf() { return leaf; }
    public void setLeaf(boolean leaf) { this.leaf = leaf; }

    private String fileName;
    public String getFileName() { return fileName; }
    public void setFileName(String fileName) { this.fileName = fileName; }

    private Object fileSize;
    public Object getFileSize() { return fileSize; }
    public void setFileSize(Object fileSize) { this.fileSize = fileSize; }

    private Date fileDate;
    public Date getFileDate() { return fileDate; }
    public void setFileDate(Date fileDate) { this.fileDate = fileDate; }

    private String filePath;
    public String getFilePath() { return filePath; }
    public void setFilePath(String filePath) { this.filePath = filePath; }

    private List<FileNode> children;
    public List<FileNode> getChildren() { return children; }
    public void setChildren(List<FileNode> children) { this.children = children; }

    @Override
    public String toString() {
        return "FileNode [leaf=" + leaf + ", fileName=" + fileName
                + ", fileSize=" + fileSize + ", filePath=" + filePath + "]";
    }
}

src/main/java/logviewer/service/LogViewerService.java
package logviewer.service;

/**
 * Created by moyong on 16/10/28.
 */

import logviewer.domain.FileNode;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import javax.annotation.PostConstruct;
import javax.annotation.Resource;
import java.io.File;
import java.io.IOException;
import java.nio.file.*;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Service
public class LogViewerService {

    protected final Logger log = LogManager.getLogger(getClass());
    private String logDir;//="/Users/moyong/docker-share/logs";
    private String logDirPathName;
    @Resource
    private Environment env;
    //@Autowired private MyProperties myProperties;

    /**
     * If the log directory changes, restart app in order to get the new property
     * into this service.
     */
    @PostConstruct  //通过@PostConstruct 和 @PreDestroy 方法 实现初始化和销毁bean之前进行的操作
    public void init() {

        System.out.println(env.getRequiredProperty("logviewer.path"));

        logDir = env.getRequiredProperty("logviewer.path");
//        logDir = myProperties.getProperty(PropertyConstants.TOMCAT_LOG_DIR);
    }

    /**
     * Get a list of all the files, including files in subdirectories.
     * @return List of all files found in the log directory, including subdirectories
     */
    public List<FileNode> getFiles()  {
        List<FileNode> fileList = new ArrayList<FileNode>();
        File dir = new File(logDir);
        logDirPathName = dir.getPath();
        listFilesInDirectory(dir, fileList);
        return fileList;
    }

    /**
     * This builds up the list of the files, including in subdirectories.  This
     * is recursive for each subdirectory found
     * @param dir The directory currently finding files in
     * @param currentDir The list to add all files found in the current directory
     */
    protected void listFilesInDirectory(File dir, List<FileNode> currentDir) {
        File[] files = dir.listFiles();
        if (files != null) {
            for (File f : files) {
                if (f.isDirectory()) {
                    List<FileNode> dirTree = new ArrayList<FileNode>();
                    FileNode node = addDirNode(f);
                    currentDir.add(node);
                    listFilesInDirectory(f, dirTree);
                    node.setChildren(dirTree);
                } else {
                    FileNode node = addFileNode(f);
                    currentDir.add(node);
                }
            }
        }
    }

    /**
     * Create a file node.  File nodes indicate they are a leaf
     * and include the file path.
     * @param diskFile The file on disk
     * @return FileNode, populated as a file node
     */
    protected FileNode addFileNode(File diskFile) {
        FileNode node = new FileNode();
        String diskFileName = diskFile.getName();
        node.setFileName(diskFileName);
        node.setFileDate(new Date(diskFile.lastModified()));
        node.setLeaf(true);
        node.setFileSize(new Long(diskFile.length()));

        String parentName = diskFile.getPath();
        String filePath = parentName.substring(logDirPathName.length() + 1);
        node.setFilePath(filePath);

        return node;
    }

    /**
     * Create a directory node.  Directory nodes do not have a file path and
     * are <i>not</i> leafs.  Their file size is set to "dir"
     * @param diskFile
     * @return FileNode, populated as a directory node
     */
    protected FileNode addDirNode(File diskFile) {
        FileNode node = new FileNode();
        String diskFileName = diskFile.getName();
        node.setFileName(diskFileName);
        node.setFileDate(new Date(diskFile.lastModified()));
        node.setFileSize("<dir>");
        return node;
    }


    /**
     * Get the file from the log directory.
     * @param fileName FileName includes path information if file is not in the root
     * log directory.
     * @return file found
     */
    public File getFile(String fileName) {
        File file = new File(logDir + "/" + fileName);
        return file;
    }

    public boolean save(MultipartFile file) {
        Path target = Paths.get(logDir,file.getOriginalFilename());
        System.out.println(target.getFileName());
        System.out.println(target.getParent());
        System.out.println(target.getRoot());
        try {
            Files.copy(file.getInputStream(), target, StandardCopyOption.REPLACE_EXISTING);
            return true;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }
}


src/main/resources/static/logviewer/app/controller/Viewport.js
Ext.define('LogViewer.controller.Viewport', {
    extend: 'Ext.app.Controller',

    requires: [
        'Ext.grid.column.Action'
    ],

    stores: [
        'Files'
    ],

    refs: [
        { ref: 'fileTree', selector: '#fileTree' },
        { ref: 'alertBox', selector: '#myalert'}
    ],

    init : function () {
        Ext.log('init viewport controller');
        this.getFilesStore().getProxy().on({
            exception: this.onProxyException,
            scope: this
        });
        this.control({
            'viewport #fileTree' : {
                itemclick : this.fileClick
            },
            'viewport tool[type="refresh"]' : {
                click: this.onRefresh
            },
            'viewport actioncolumn' : {
                click: this.handleActionColumn
            }
        });
    },

    fileClick : function(view, record, item, index, event) {
        Ext.log("in fileClick ");
        var logWin, fileName, cleanName, logUrl;

        if (record.isLeaf()) {
            fileName = record.get('filePath');
            logUrl = Ext.String.urlAppend('/logviewer/open', 'fileName='+fileName);
            cleanName = fileName.replace(/\W/g, '')
            Ext.log('open [' + logUrl + '] name [' + cleanName + ']');
            // window.location = logUrl;

            //window.open(logUrl,'_blank');
            var dmnurl = Ext.String.urlAppend('/dmn/index.html', 'dmnFileName='+encodeURIComponent(logUrl));
            window.open(dmnurl,'_blank');


            //是先到后台将要导出或者下载的数据 准备后，并将文件路径传到 前台，然后再指向该路径
            // Ext.Ajax.request({
            //     url:logUrl,
            //     success:function(res){
            //         var obj =Ext.decode(res.responseText);
            //         window.location.href =obj.path;
            //     }
            // });
        }
    },

    onRefresh : function() {
        Ext.log('in refresh');
        var fileTree = this.getFileTree();
        this.getAlertBox().hide();
        fileTree.getStore().load();
    },

    handleActionColumn : function(view, row, col, item, e, record) {
        var logWin, fileName, downloadUrl;
        Ext.log("handleActionColumn");
        if (record.isLeaf()) {
            fileName = record.get('filePath');
            this.onDownload(fileName);
        }
        return false;
    },

    onDownload : function(fileName) {
        Ext.log("Exporting to file:"+fileName);

        var frame, form, hidden, params, url;

        //frame = Ext.fly('exportframe').dom;
        //frame.src = Ext.SSL_SECURE_URL;

        form = Ext.fly('exportform').dom;

        //url = Ext.String.urlAppend('/logviewer/download', 'fileName='+fileName);
        //cleanName = fileName.replace(/\W/g, '')
        //Ext.log('download [' + url + '] name [' + cleanName + ']');
        //url = '../util/htmlresponse.json';
        //Ext.log(url)

        form.action = '/logviewer/download';

        //数据要么走 url 要么走 hidden
        hidden = document.getElementById('excelconfig');
        params = {fileName: fileName};
        hidden.value = Ext.encode(fileName);
        // Ext.log(fileName);

        // hidden.value = fileName;

        form.submit();
    },

    onProxyException : function(proxy, response, operation) {
        var alertBox = this.getAlertBox();
        alertBox.showError("An unexpected error occurred.");
    }
});

src/main/java/org/camunda/bpm/spring/boot/example/dmn/rest/DmnRestApplication.java
package org.camunda.bpm.spring.boot.example.dmn.rest;

import org.camunda.bpm.engine.ProcessEngine;
import org.camunda.bpm.engine.impl.cfg.AbstractProcessEnginePlugin;
import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl;
import org.camunda.bpm.engine.impl.cfg.ProcessEnginePlugin;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
//import org.springframework.web.servlet.config.annotation.EnableWebMvc;

@SpringBootApplication
//@EnableWebMvc
@ComponentScan({"logviewer"})
public class DmnRestApplication {

  public static void main(final String... args) throws Exception {
    SpringApplication.run(DmnRestApplication.class, args);
  }

  private final Logger logger = LoggerFactory.getLogger(this.getClass());

  @Bean
  public ProcessEnginePlugin statusPlugin() {
    return new AbstractProcessEnginePlugin() {
      @Override
      public void postInit(ProcessEngineConfigurationImpl processEngineConfiguration) {
        logger.warn("post init");
      }

      @Override
      public void postProcessEngineBuild(ProcessEngine processEngine) {
        logger.warn("post build");
      }

      @Override
      public void preInit(ProcessEngineConfigurationImpl processEngineConfiguration) {
        logger.warn("pre init");
      }
    };
  }

}


src/main/java/logviewer/web/LogViewerController.java
package logviewer.web;

/**
 * Created by moyong on 16/10/28.
 */


import logviewer.domain.FileNode;
import logviewer.service.LogViewerService;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.util.FileCopyUtils;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.nio.file.Files;
import java.nio.file.LinkOption;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

@Controller
@RequestMapping("/logviewer")
public class LogViewerController {

    protected final Logger log = LogManager.getLogger(getClass());
    @Autowired
    private LogViewerService logViewerService;

    /**
     * Build list of all files in the log directory
     *
     * @return JSON response with objects expected by a tree grid.
     */
    @RequestMapping(value = "/files")
    public
    @ResponseBody
    Object getFiles(HttpServletResponse response) {
        try {
            List<FileNode> files = logViewerService.getFiles();
            return files;
        } catch (Exception e) {
            log.error("Error loading files ", e);
            // This just returns a Map that will result in JSON {"success":"false", "message":"Unexpected Error"}
//            Map<String, Object> errorMap = WebUtil.getModelMapMessage(false, ControllerConstants.DEFAULT_ERROR_MSG, null);
            Map<String, Object> errorMap = new HashMap();
            errorMap.put("error", e.getMessage());
            return errorMap;
        }
    }

    /**
     * Open a specified file in a browser window/tab.  Some of the success of the
     * open might depend on client-side settings.
     *
     * @param request
     * @param response
     * @param fileName The name of the file to open, including path info if not in
     *                 the root log directory.
     */
    @RequestMapping(value = "/open")
    public void openFile(HttpServletRequest request, HttpServletResponse response,
                         @RequestParam("fileName") String fileName) {
        try {

            File file = logViewerService.getFile(fileName);


            byte[] content = new byte[(int) file.length()];
            response.setContentType("text/plain");
            response.addHeader("content-disposition", "inline;filename=" + fileName);
            response.setContentLength(content.length);
            FileInputStream in = new FileInputStream(file);
            FileCopyUtils.copy(in, response.getOutputStream());
        } catch (Exception e) {
            log.error("Something happened", e);
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }
    }

    /**
     * Download the selected file to the client.
     *
     * @param request
     * @param response
     * @param fileName The name of the file to open, including path info if not in
     *                 the root log directory.
     */
    @RequestMapping(value = "/download", method = RequestMethod.POST)
    public void downloadFile(HttpServletRequest request, HttpServletResponse response,
                             @RequestParam("fileName") String fileName) {
        try {

            System.out.println("============");
            System.out.println(fileName);
            System.out.println("============");
            String[] name=fileName.split(",");


            File file = logViewerService.getFile(name[0].replace("\"",""));

            log.debug(file.getName());
            log.debug(file.exists());

            byte[] content = new byte[(int) file.length()];
            response.setContentType("application/octet-stream");
            response.setHeader("Content-Disposition", "attachment;filename=" + fileName);
            response.setHeader("Cache-Control", "cache, must-revalidate");
            response.setHeader("Pragma", "public");
            response.setHeader("Content-Transfer-Encoding", "binary");
            response.setContentLength(content.length);
            FileInputStream in = new FileInputStream(file);
            FileCopyUtils.copy(in, response.getOutputStream());

        } catch (Exception e) {
            log.error("Something happened", e);
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }
    }


    /**
     * 单个文件上传，测试通过
     * @param file
     * @param desc
     * @return
     */
    @RequestMapping(value="/upload", method=RequestMethod.POST )
    public @ResponseBody
    Map singleSave(@RequestParam("upload") MultipartFile file){
//        System.out.println("File Description:"+desc);
        Map result = new HashMap();

        if (!file.isEmpty()) {
//            try {

            log.debug(file.getOriginalFilename());
            System.out.println(file.getContentType());
            System.out.println(file.getName());
            System.out.println(file.getOriginalFilename());

            boolean save = logViewerService.save(file);


//
//                String filename = UUID.randomUUID().toString()+"."+FileUtils.getExtensionName(file.getOriginalFilename());
//
//                Path target = Paths.get(ROOT, filename);
//                Files.copy(file.getInputStream(), target);
//
//                log.debug(target.getFileName().toString());
//                log.debug(target.getParent().getFileName().toString());
//                log.debug(target.toString());
//                log.debug(target.toRealPath(LinkOption.NOFOLLOW_LINKS).toString());
//
//                log.debug("=============ContentType:"+file.getContentType());
////                FilenameUtils.getExtension()
//                log.debug(Files.probeContentType(target));
//
//
//                Avatar img=new Avatar();
//                img.setFilename(file.getOriginalFilename());
//                img.setFilesize(file.getSize());
//                img.setWebpath(ROOT+"/"+filename);
//                img.setSyspath(target.toRealPath(LinkOption.NOFOLLOW_LINKS).toString());
//                img.setCreateDate(new Date());
//                img.setCreateBy("user");
//                avatarRepository.save(img);
//
//
//                result.put("data","");
//                result.put("upload",img);
//                Map m = new HashMap();
//
//                List<Avatar> files = avatarRepository.findAll();
//                Map mm=new HashMap();
//                for (Avatar f:files
//                        ) {
//                    mm.put(f.getId(),f);
//                }
//                m.put("files", mm);
//
//                result.put("files",m);
////                result.put("files",avatarRepository.findAll());
//
            result.put("desc","文件上传成功"+file.getOriginalFilename());
            result.put("success",save);

                return result;
//
//            } catch (IOException e) {
//                result.put("desc","文件上传失败"+file.getOriginalFilename());
//                return result;
//            }
        } else {
            result.put("desc", "Failed to upload " + file.getOriginalFilename() + " because it was empty");
            return result;
        }

    }
}
